### PR TITLE
[LF] fix versioning of value w.r.t. interface

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala [hunk 1]
```diff
@@ -105,5 +104,4 @@ final case class InterfaceView(
     templateId: Identifier,
     argument: SValue,
     interfaceId: Identifier,
-    interfaceVersion: TransactionVersion,
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:108]: Speedy commands/values are not versioned. No reason to put one here.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 2]
```diff
@@ -1023,7 +1024,10 @@ private[lf] object SBuiltin {
         machine: Machine,
         onLedger: OnLedger,
     ): Control = {
-      val chosenValue = machine.normValue(templateId, args.get(0))
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1026]: This was wrongly versioned.
I will add a test before merging.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 3]
```diff
@@ -351,9 +351,6 @@ private[lf] object Speedy {
         compiledPackages.pkgInterface.packageLanguageVersion(tmplId.packageId)
       )
 
-    def normValue(templateId: TypeConName, svalue: SValue): V =
-      svalue.toNormalizedValue(tmplId2TxVersion(templateId))
-
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:356]: This was wrongly used. Better to deleted it, and force people to compute the version explicitly before calling `toNormalizedValue`. 
Furthermore it induced most of the time unnecessary recomputation of `tmplId2TxVersion(templateId)`

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 4]
```diff
@@ -1023,22 +1027,22 @@ private[lf] object SBuiltin {
...
       val templateObservers = cached.observers
       val ctrls = extractParties(NameOf.qualifiedNameOfCurrentFunc, args.get(2))
       onLedger.enforceChoiceControllersLimit(ctrls, coid, templateId, choiceId, chosenValue)
       val obsrs = extractParties(NameOf.qualifiedNameOfCurrentFunc, args.get(3))
       onLedger.enforceChoiceObserversLimit(obsrs, coid, templateId, choiceId, chosenValue)
-      val mbKey = cached.key
-      val templateVersion = machine.tmplId2TxVersion(templateId)
-      val interfaceVersion = interfaceId.map(machine.tmplId2TxVersion)
-      val exerciseVersion = interfaceVersion.fold(templateVersion)(_.max(templateVersion))
+      val mbKey = cached.key.map(_.toNormalizedKeyWithMaintainers(exerciseVersion))
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1045]: This was wrongly versioned as well.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala (hunk 1 above)
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:108]: Speedy commands/values are not versioned. No reason to put one here.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala [hunk 5]
```diff
@@ -140,9 +139,6 @@ class IdeLedgerClient(
       arg: Value,
   ): Value = {
 
-    // TODO https://github.com/digital-asset/daml/issues/14830
```
Comment (nickchapman-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala:143]: Nice!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala (hunk 1 above)
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:108]: :+1:

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 6]
```diff
@@ -102,12 +102,19 @@ private[lf] object Speedy {
...
+    val unnormalizedKeyWithMaintainers = Node.KeyWithMaintainers(unnormalizedKeyValue, maintainers)
+  }
+
   private[lf] final case class CachedContract(
       templateId: Ref.TypeConName,
       value: SValue,
       signatories: Set[Party],
       observers: Set[Party],
-      key: Option[Node.KeyWithMaintainers],
+      key: Option[SKeyWithMaintainers],
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:117]: Ah, I see. We had a speedy-level value-structure `CachedContract` containing a _normalized_ value-structure `Node.KeyWithMainters` directly.

And the fix is to introduce and use a corresponding speedy-level type `SKeyWithMaintainers`.

Nice!

**Review by nickchapman-da — APPROVED** [review]
:+1: LGTM

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:117]: Indeed.

**Pull request merged**